lib/ipaddr.mli

```diff
@@ -148,6 +148,14 @@ module V4 : sig
       suffix, and an IPv4 address prefixed. *)
   val of_domain_name : 'a Domain_name.t -> t option
 
+  (** {3 Utility functions} *)
+
+  (** [succ ipv4] is ip address next to [ipv4] *)
```

Comment:
Could we specify here the behaviour of the function if it is passed the maximum ipv4 value (i.e. wraparound or does it return the same value)

Reply:
Oh, I now see inconsistency with V4.succ and V6.succ - V4 relies on Int32 behaviour that cycles(255.255.255.255 -> 0.0.0.0) while my implementation for B128 raises exception (failwith "B128 overflow"). Which behaviour should we support here?
- cycle
- exception
- do nothing and return input value
- something other

Reply:
@avsm which behavior should I choose here?

Reply:
I'm against an exception (since ipaddr 4.0.0 (mostly) got rid of exceptions in the public interface), and would be in favour to return a `t option` -- which is explicit and requires the caller to handle this.

Reply:
So, we should not silently overflow int type? Is it appropriate to raise exception for unexpected code path (e.g. `    | _ -> failwith "Unexpected error with B128"`)?

Reply:
I find it not appropriate to raise exceptions. But then, I also don't quite understand who'll use such a function `succ`.

Reply:
For example IPAM or DHCP server implemented in OCaml.

Reply:
For V4 `succ` is just `let succ = Int32.succ`, so it inherits behavior of stdlib's Int32 module. It silently overflows. Int32 docs says: `Int32.succ x is Int32.add x Int32.one.`. Also, I want to note that ipaddr library already has `(t, [> `Msg of string ]) result` type for errors. Error handling is a holywar topic, I'll just implement any decision you approve, but I need this approvement. :)

Reply:
Let's go with being very explicit and returning a `(t, [>Msg of string)` return type for `succ`.  The users of this will want to know when it is overflowing -- for instance, a DHCP server that runs to the end of the IP allocation pool will not want to roll over to the beginning of the range.

Reply:
Wonder if a `val Prefix.succ : Prefix.t -> t -> (t, [> `Msg of string]) result` would make sense as well, useful for enumerating within a range defined by a given prefix?

Reply:
You mean `succ` and `pred` with a bounds check? It can be composed with `succ` and `Prefix.mem`, but would be useful.

Reply:
Yes, I just thought that `succ` with a bounds check for overflow (checking that it's within `/0`) could be generalized to `/1`, `/24` etc as well, and that it would be useful for those things too, while we are at it. :-)